File: examples/favorites/typescript/favorites.ts
import {
  Account,
  Pubkey,
  Result,
  u64,
  Signer,
  Vec,
  Str,
} from "@solanaturbine/poseidon";

export default class FavoritesProgram {
  static PROGRAM_ID = new Pubkey("11111111111111111111111111111111");

  setFavorites(
    owner: Signer,
    number: u64,
    color: Str<50>,
    hobbies: Vec<Str<50>, 5>,
    favorites: Favorites,
  ): Result {
    favorites.derive(["favorites", owner.key]).initIfNeeded(owner);

    favorites.number = number;
    favorites.color = color;
    favorites.hobbies = hobbies;
  }
}

export interface Favorites extends Account {
  number: u64;
  color: Str<50>;
  hobbies: Vec<Str<50>, 5>;
}

File: examples/favorites/rust/favorites.rs
use anchor_lang::prelude::*;
declare_id!("11111111111111111111111111111111");
#[program]
pub mod favorites_program {
    use super::*;
    pub fn set_favorites(
        ctx: Context<SetFavoritesContext>,
        number: u64,
        color: String,
        hobbies: Vec<String>,
    ) -> Result<()> {
        ctx.accounts.favorites.number = number;
        ctx.accounts.favorites.color = color;
        ctx.accounts.favorites.hobbies = hobbies;
        Ok(())
    }
}
#[derive(Accounts)]
pub struct SetFavoritesContext<'info> {
    #[account(
        init_if_needed,
        payer = owner,
        space = 344,
        seeds = [b"favorites",
        owner.key().as_ref()],
        bump,
    )]
    pub favorites: Account<'info, Favorites>,
    #[account(mut)]
    pub owner: Signer<'info>,
    pub system_program: Program<'info, System>,
}
#[account]
pub struct Favorites {
    pub number: u64,
    pub color: String,
    pub hobbies: Vec<String>,
}
